### ISSUE TITLE
num_batches_per_epoch not well accounted for by gluonts.torch.model.deepar.DeepAREstimator

### ISSUE THREAD

**Issue by pbruneau** [awslabs/gluon-ts#2594]
## Description
During each epoch, validation starts at the `num_batches_per_epoch` specified in the `DeepAREstimator` constructor call, but after this, training exhausts the whole training dataset, while validation runs concurrently.

Maybe I am not using the API correctly, or something is wrong with the gluonts version I'm using?

## To Reproduce
Please check out [this minimal example](https://gist.github.com/pbruneau/bef11c78662a2a2dbfcf6381be4f2c04)

## Error message or code output
Before `num_batches_per_epoch` is reached, training proceeds as expected:
```
141 K     Trainable params
0         Non-trainable params
141 K     Total params
0.564     Total estimated model params size (MB)
Epoch 0: : 24it [00:01, 23.54it/s, loss=7.04, v_num=35]
```

After it is reached (50 in that case), training and validation run concurrently:
```
141 K     Trainable params
0         Non-trainable params
141 K     Total params
0.564     Total estimated model params size (MB)
Epoch 0: : 169it [00:08, 19.96it/s, loss=6.34, v_num=37]                                                                                                                        
Validation DataLoader 0: : 119it [00:05, 20.99it/s]
```

## Environment
- Operating system: Linux
- Python version: 3.7.10
- GluonTS version: v0.11.7
- MXNet version: 1.9.1

**Comment by lostella** [comment]
@pbruneau I simplified your example by removing `num_workers`:

```python
from gluonts.dataset.repository.datasets import get_dataset
from gluonts.torch.model.deepar import DeepAREstimator

dataset = get_dataset("electricity")
context_length = 2 * 7 * 24
prediction_length = dataset.metadata.prediction_length

model = DeepAREstimator(
    prediction_length=prediction_length,
    context_length=context_length,
    freq='1H',
    batch_size=32,
    num_batches_per_epoch=2,
)

predictor = model.train(
    dataset.train,
    dataset.test,
)
```

The validation data consists of 2247 series in this case, which are 71 batches of size 32; after the `num_batches_per_epoch=2` batches are consumed for training, 71 are accounted for in the validation logging line, adding up to 73. I think the first logging line is just counting and timing the total number of batches processed.

It's left to see whether
* logging is working as expected (this is handled by PyTorch Lightning, but the estimator class may be using it wrong)
* `num_workers` is working as expected

**Comment by pbruneau** [comment]
Thanks for the follow-up @lostella, this is not the expected behaviour but at least you explain the numbers!

Following up on this, I tried to run the same example, just adding `num_workers=2`
The number of training steps looks right (i.e. validation starts after `num_batches_per_epoch` steps), but the number of validation steps is doubled... Maybe something with the slight differences between `create_training_data_loader` and `create_training_data_loader`? Looking at the code, I did not see anything obvious...

**Comment by lostella** [comment]
> this is not the expected behaviour but at least you explain the numbers!

You mean logging should not present information that way?

In the `num_workers=2` case, it seems like the two worker processes are consuming the whole set of validation batches each? If so, this is something that needs to be looked into

**Comment by pbruneau** [comment]
> You mean logging should not present information that way?

Well, I would expect it to behave like the default gluonts progress bar (and like the progress bars with TF, for example), with distinct progress for train and validation. But I agree this is minor (and possibly tied to PyTorch Lightning as you mentioned).

> In the num_workers=2 case, it seems like the two worker processes are consuming the whole set of validation batches each? If so, this is something that needs to be looked into

Maybe so, just that the count is aggregated in a single bar,
that is, with num_workers=2, the validation counter goes up to 142. Behind the scenes, indeed this could be that each worker consumes the whole validation set.

**Comment by gorold** [comment]
The issue of worker processes independently consuming the entire dataset affects not only validation, but also training set as well. This seems to be well documented [here](https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset). While #2615 solves this issue, in my experiments where data loading is the major bottleneck, this impacts performance severely. Another more performant fix is to implement this selection logic in the underlying map-style gluonts dataset (arrow table, dictionary, etc.).

**Comment by lostella** [comment]
Yes, I think #2615 has the drawback that workers will still process the elements they skip (so all lazily evaluated transformation will still take time and slow the worker down)

**Comment by Alex2251** [comment]
> The issue of worker processes independently consuming the entire dataset affects not only validation, but also training set as well. This seems to be well documented [here](https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset). While #2615 solves this issue, in my experiments where data loading is the major bottleneck, this impacts performance severely. Another more performant fix is to implement this selection logic in the underlying map-style gluonts dataset (arrow table, dictionary, etc.).

Yes, you right. It better to use  map-style if it possible but if you use ItarableDataset class it means you can't. But the sistem should work predictable. Because today I have 52 000 timeseries and 10 workers and batch is 100 elements, approximatly one epoch takes 520 unique bathes and 52 iterations but in practice I will cover all data (one epoc) in 520 iterations it looks dramatikal and unpredictable. It can be better to add shuffle for train and presplit for validation and test but it's still difficult for streaming data. This commit solves problem in general and another solution can be done for specific tasks.

**Comment by gorold** [comment]
Agreed, I do feel that #2615 should be rolled out as a quick fix, since this is essentially quite a critical bug -- if you use 10 workers, you will see 10 of the same batches in a row during training. A more performant fix for map datasets can be made later.

**Comment by jaheba** [comment]
I think this can be closed by #2628.

@pbruneau Can you verify if the behaviour is as expected on `dev`?

**Comment by pbruneau** [comment]
I have just checked, the numbers displayed look right, thanks! I'm closing the issue then.

I have a subsidiary question though; considering the example code from @lostella above:
```
from gluonts.dataset.repository.datasets import get_dataset
from gluonts.torch.model.deepar import DeepAREstimator

dataset = get_dataset("electricity")
context_length = 2 * 7 * 24
prediction_length = dataset.metadata.prediction_length

model = DeepAREstimator(
    prediction_length=prediction_length,
    context_length=context_length,
    freq='1H',
    batch_size=32,
    num_batches_per_epoch=2,
)

predictor = model.train(
    dataset.train,
    dataset.test,
)
```

If I change `num_workers`  in the `model.train` call, training seems to be faster indeed according to my monitor. But if I try to use my GPU with these lines juste before `model.train`:
```
model.trainer_kwargs={
    'gpus': -1 if torch.cuda.is_available() else None,
}
```
Having `num_workers > 1` does now not seem to bring any benefit. Having several workers should be the way to increase GPU occupancy, right? Maybe there is an additional step to effectively do so?

**Comment by jaheba** [comment]
We've removed `num_workers` from the `train`, so my assumption is that it should have no effect. We probably should add a warning if arguments are passed that are not used.

We want to re-add parallel dataloading in the future, but do it correctly that time. There is some work to be done to shard the underlying data correctly first.

**Comment by pbruneau** [comment]
All right, in my old version `num_workers` made a difference, but indeed since I pulled in `dev` it is not the case any more. Thanks for the anwser @jaheba!
